src/vs/editor/common/modes.ts

```diff
@@ -284,7 +292,7 @@ export interface CodeActionProvider {
 	/**
 	 * Provide commands for the given document and range.
 	 */
-	provideCodeActions(model: editorCommon.IReadOnlyModel, range: Range, token: CancellationToken): Command[] | Thenable<Command[]>;
+	provideCodeActions(model: editorCommon.IReadOnlyModel, range: Range, token: CancellationToken): (Command | CodeAction)[] | Thenable<(Command | CodeAction)[]>;
```

Comment:
In this land we can break the API so I'd suggest we don't carry along the `Command | CodeAction` notion but just make it a `CodeAction`. I'd also do that on the protocol layer and do the `Command` becomes `CodeAction` thing  in the code action adapter.

Reply:
I'm concerned this will break the `vscode.executeCodeActionProvider` command. I believe this should still return commands instead of code actions for old providers and so  I'm going to revert the change to only use code actions for now